slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -279,7 +279,81 @@ ::grpc::Status SLAMServiceImpl::GetCurrentPointCloudMap(
 ::grpc::Status SLAMServiceImpl::GetPointCloudMapStream(
     ServerContext *context, const GetPointCloudMapStreamRequest *request,
     ServerWriter<GetPointCloudMapStreamResponse> *writer) {
-    return ::grpc::Status(::grpc::StatusCode::UNIMPLEMENTED, "");
+    std::string pointcloud_map;
+    int num_points;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            num_points = GetLatestSampledPointCloudMapString(pointcloud_map);
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_pointcloud_map = pointcloud_map;
+            latest_num_points = num_points;
+        } else {
+            // Either we are in localization mode or we couldn't lock the mutex
+            // which means the optimization process locked it and we need to use
+            // the backed up latest map
+            if (action_mode == ActionMode::LOCALIZING) {
+                LOG(INFO)
+                    << "In localization mode, using cached pointcloud map";
+            } else {
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+            num_points = latest_num_points;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "map pointcloud does not have points yet");
+    }
+
+    std::string pcd_chunk;
+    GetPointCloudMapStreamResponse response;
+
+    // Calculate chunk sizes to ensure no points data is split between chunks
+    int header_byte_size = viam::utils::pcdHeader(num_points, true).size();
+    int point_byte_size = 4 * sizeof(float);  // X, Y, Z, Color
```

Comment:
https://github.com/viamrobotics/slam/pull/150/files#diff-f582be4089c313a31e0dfa10d48874bfa84a0fc08a2983d6ceea52b161e574d6R680

I think this should be `3 * sizeof(float) + sizeof(int);` given that the last value is an int, not a float.